spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializer.scala

```diff
@@ -0,0 +1,207 @@
+        val ptype = param.typeSignature
+        if (ptype.toString == "scala.reflect.ClassTag[T]") {
+          args(i) = evidence
+        } else if (ptype.toString ==
+          tensorNumericType.toString) {
+          args(i) = ev
+        } else {
+          require(modelAttributes.containsKey(name), s"$name value cannot be found")
+          val attribute = modelAttributes.get(name)
+          val value = DataConverter.getAttributeValue(attribute)
+          args(i) = value
+        }
+        i+= 1
+      })
+    })
+    val module = constructorMirror.apply(args : _*).
+      asInstanceOf[AbstractModule[Activity, Activity, T]]
+    createBigDLModule(model, module)
+  }
+
+  override def serializeModule[T: ClassTag](module : ModuleData[T])
+                                           (implicit ev: TensorNumeric[T]) : BigDLModule = {
+    val bigDLModelBuilder = BigDLModule.newBuilder
+    val cls = module.module.getClass
+    val moduleType = getModuleTypeByCls(cls)
+    bigDLModelBuilder.setModuleType(moduleType)
+    val fullParams = getCostructorMirror(cls).symbol.paramss
+    val clsTag = scala.reflect.classTag[T]
+    val constructorParams = fullParams(0)
+    constructorParams.foreach(param => {
+      val paramName = param.name.decodedName.toString
+      var ptype = param.typeSignature
+      val attrBuilder = AttrValue.newBuilder
+      // For some modules, fields are declared inside but passed to Super directly
+      var field : Field = null
+      try {
+        field = cls.getDeclaredField(paramName)
+      } catch {
+        case e : NoSuchFieldException =>
+          field = cls.getSuperclass.getDeclaredField(paramName)
+      }
+      field.setAccessible(true)
+      val fieldValue = field.get(module.module)
+      DataConverter.setAttributeValue(attrBuilder, fieldValue, ptype)
+      bigDLModelBuilder.putAttr(paramName, attrBuilder.build)
+    })
+    createSerializeBigDLModule(bigDLModelBuilder, module)
+  }
+
+
+  def serialize[T: ClassTag](bigDLModule : ModuleData[T])
+                            (implicit ev: TensorNumeric[T])
+    : BigDLModule = {
+    val module = bigDLModule.module
+    val cls = module.getClass
+    serializerMaps(cls).serializeModule(bigDLModule)
+  }
+
+  def load[T: ClassTag](model: BigDLModule)
+                       (implicit ev: TensorNumeric[T]) : ModuleData[T] = {
+    deserializerMaps(model.getModuleType).loadModule(model)
+  }
+
+
+
+  def registerModule(moduleType : String, moduleCls : Class[_],
+    serializer : ModuleSerializable) : Unit = {
+    moduleMaps(moduleType) = moduleCls
+    classMaps(moduleCls) = moduleType
+    serializerMaps(moduleCls) = serializer
+    deserializerMaps(moduleType) = serializer
+  }
+
+  def getModuleClsByType(moduleType : String) : Class[_] = {
+    require(moduleMaps.contains(moduleType), s"$moduleType is not supported")
+    moduleMaps(moduleType)
+  }
+
+  def getModuleTypeByCls(cls : Class[_]) : String = {
```

Comment:
all public method should have java doc. Looks like it is an utility, should it be private or private[nn]?